src/main/java/org/broadinstitute/hellbender/utils/clipping/ClippingOp.java

```diff
@@ -599,24 +597,94 @@ private int calcHardSoftOffset(final Cigar cigar) {
+        while (i < elements.size() && elements.get(i).getOperator() == CigarOperator.SOFT_CLIP) {
             size += elements.get(i).getLength();
             i++;
         }
 
         return size;
     }
 
+    /**
+    * Calculates shift of alignment int the newCigar relative to the oldCigar due to
+     * hard/soft clipping under assumption that the
+    * original clipped bases did not contain insertions/deletions. This is a naive code that
+    * does not work for many CIGAR strings
+     *
+     * @param oldCigar original cigar
+     * @param newCigar new cigar (with hard/soft clipping)
+     * @return int the offset (from 0 between the alignment start on the old and on the new cigar)
+    */
     private int calculateAlignmentStartShift(final Cigar oldCigar, final Cigar newCigar) {
         final int newShift = calcHardSoftOffset(newCigar);
         final int oldShift = calcHardSoftOffset(oldCigar);
         return newShift - oldShift;
     }
 
+    /**
+    /* Calculates how much the alignment should be shifted when hard/soft clipping is applied
+    * to the cigar
+     * @param oldCigar the original CIGAR
+     * @param clipping - number of bases of the read clipped
+     * @return int - the offset between the alignment starts in the oldCigar and in
+     * the cigar after applying clipping
+     */
+    private int calculateAlignmentStartShift(final Cigar oldCigar, final int clipping) {
+
+        final int newReadBasesClipped = clipping;
+        int readBasesClipped = 0; // The number of read bases consumed on the new cigar before reference bases are consumed
+        int refBasesClipped = 0; // A measure of the reference offset between the oldCigar and the clippedCigar
+
+        boolean final_element = false;
+        for (final CigarElement e : oldCigar.getCigarElements()) {
+
+
+            int curRefLength = e.getLength();
+            int curReadLength = e.getLength();
+
+            // needed only if the clipping ended on N or D
+            if (final_element) {
+                if (!e.getOperator().consumesReadBases() && e.getOperator().consumesReferenceBases()) {
+                    refBasesClipped += curRefLength;
+                }
+                break;
+            }
+
+            if (!e.getOperator().consumesReadBases()) {
+                curReadLength = 0;
+            }
+
+            final boolean truncated;
+            if (readBasesClipped + curReadLength > newReadBasesClipped) {
+                curReadLength = newReadBasesClipped - readBasesClipped;
+                curRefLength = curReadLength;
+                truncated = true;
+            } else {
+                truncated= false;
+            }
+
+            if (!e.getOperator().consumesReferenceBases()) {
+                curRefLength = 0;
+            }
+
+            readBasesClipped += curReadLength;
+            refBasesClipped += curRefLength;
+
+            if (readBasesClipped > newReadBasesClipped || truncated) {
+                break;
```

Comment:
The control flow is tricky to follow here because there are two fairly distant `breaks` as well as a `final_element = true` that then triggers a `break` partway through the next iteration.  Can this be made clearer?

Reply:
yes...it's tricky....not sure how to make it less hard to follow.

Reply:
in resolving this I found a bug! thanks....will resubmit soon.